gene/mutate.go

```diff
@@ -167,59 +174,87 @@ func Insertion(g *Gene, genomePos int, alt []dna.Base) (EffectPrediction, error)
-		//var currCodon dna.Codon
-		//
-		//if frame != 0 { // if insertion disrupts a codon, note the codon being changed
-		//	currCodon, err = CdnaPosToCodon(g, cdnaPos)
-		//	answer.AaRef = []dna.AminoAcid{dna.TranslateCodon(&currCodon)}
-		//}
+		answer.CdnaPos, answer.CdnaDist, err = GenomicPosToCdna(g, genomePos+1)
+		answer.CdnaPos-- // Subtract 1 to get base zero
+		frame := (cdnaPos + 1) % 3
+		var currCodon dna.Codon
+
+		if frame != 0 { // if insertion disrupts a codon, note the codon being changed
+			currCodon, err = CdnaPosToCodon(g, cdnaPos)
+			answer.AaRef = []dna.AminoAcid{dna.TranslateCodon(&currCodon)}
+		}
 
 		// Update cDNA
 		g.cdnaSeq = dna.Insert(g.cdnaSeq, int64(cdnaPos+1), alt)
 
-		//TODO: Effect Prediction WIP
-		//answer.AaPos = cdnaPos / 3
-		//
-		//if len(alt) % 3 != 0 { // Causes Frameshift // TODO this is a bit jury rigged, but it works....
-		//	answer.Consequence = Frameshift
-		//	currCodon, err = CdnaPosToCodon(g, cdnaPos + 1)
-		//	var stopFound bool = true
-		//	for currCdnaPos := cdnaPos + 1 + 3; dna.TranslateCodon(&currCodon) != dna.Stop; currCdnaPos += 3 {
-		//		answer.AaAlt = append(answer.AaAlt, dna.TranslateCodon(&currCodon))
-		//		if currCdnaPos + (3 - frame) > len(g.cdnaSeq) { // if do dont hit a stop by the end of the cdnaSeq, then move to genomic DNA
-		//			currGdnaPos := g.cdsEnds[len(g.cdsEnds)-1] + 1 - frame
-		//			for ;dna.TranslateSeq(g.genomeSeq[currGdnaPos:currGdnaPos + 3])[0] != dna.Stop; currGdnaPos += 3 {
-		//				answer.AaAlt = append(answer.AaAlt, dna.TranslateSeq(g.genomeSeq[currGdnaPos:currGdnaPos + 3])[0])
-		//				if currGdnaPos + 3 > len(g.genomeSeq) {
-		//					stopFound = false
-		//					err = errors.New("frameshift was unable to find a stop before the end of the gene")
-		//					break
-		//				}
-		//			}
-		//			break
-		//		} else {
-		//			currCodon, err = CdnaPosToCodon(g, currCdnaPos)
-		//		}
-		//	}
-		//	if stopFound {
-		//		answer.AaAlt = append(answer.AaAlt, dna.Stop)
-		//	}
-		//} else { // In-Frame
-		//	answer.Consequence = InFrameInsertion
-		//	if frame != 0 { // Disrupts an existing codon
-		//		answer.AaAlt = dna.TranslateSeq(g.cdnaSeq[(cdnaPos + 1) - frame : (cdnaPos + 1) + len(alt) + (3 - frame)])
-		//		if answer.AaRef[0] == answer.AaAlt[0] {
-		//			answer.AaRef = nil
-		//			answer.AaAlt = answer.AaAlt[1:]
-		//			answer.AaPos++
-		//		}
-		//	} else { // Does not disrupt existing codons AND is inframe
-		//		answer.AaAlt = dna.TranslateSeq(alt)
-		//	}
-		//}
+		answer.AaPos = cdnaPos / 3
+
+		if len(alt)%3 != 0 { // Causes Frameshift // TODO this is a bit jury rigged, but it works.
+			answer.StopDist = 0
+			answer.Consequence = Frameshift
+			currCodon, err = CdnaPosToCodon(g, cdnaPos+1)
+			//var stopFound bool = true
+			answer.AaAlt = dna.TranslateSeq(g.cdnaSeq[(cdnaPos+1)-frame : (cdnaPos+1)+len(alt)+(3-frame)])
+			var tmpCodon dna.Codon
+
+			for currCdnaPos := cdnaPos + 1 + 3; dna.TranslateCodon(&currCodon) != dna.Stop; currCdnaPos += 3 {
```

Comment:
I am not sure how solid the rules are for splicing, but it passed through my head that an insertion could also insert a splice site to end the exon as well as a stop codon.  Kind of an odd case and I don't think we should deal with it, but just making the comment since it crossed my mind.

Reply:
Yeah this does happen. Along the same lines you often get point mutations that activate "cryptic" splice sites which prematurely terminate the exon. These are not very straightforward to detect, and even when they do happen you often get a mixture of the normally spliced transcript and the aberrantly spliced transcript. One day this functionality would be nice to have (especially since the only good web-based program for finding these 'Human Splice Finder' was recently sold to a for-profit company) but it is beyond the scope of this pull request.